Enable Workspace-Scoped Tasks

colin-grant-work opened the pull request:
<!--
Thank you for your Pull Request. Please provide a description and review
the requirements below.

Contributors guide: https://github.com/theia-ide/theia/blob/master/CONTRIBUTING.md
-->

#### What it does
<!-- Include relevant issues and describe how they are addressed. -->
Fixes #8519
Fixes #8991 (thanks for noting that, @RomanNikitenko)
This PR sets up the necessary infrastructure to read and write tasks to a workspace file. In particular it:
 - modifies the way a `WorkspaceFilePreferenceProvider` reads and writes non-settings preferences. (breaking changes to bring Theia's workspace file schema into harmony with VSCode's)
 - modifies helper methods in `workspace-service.ts` for writing workspace files to accommodate fields other than `folders` and `settings` (breaking changes for anyone calling those file-builder methods)
 - removes some explicit references to the preference system in the workspace package.
 - adds a `WorkspaceSchemaUpdater` service to allow packages that contribute extensions to the schema of the workspace file (tasks, debug) to register their extensions
 - injects workspace scope related functionality into the `TaskConfigurationManager`

#### How to test
<!-- Explain how a reviewer can reproduce a bug, test new functionality or verify performance improvements. -->
1. Add a properly formatted `tasks` field to a workspace file. (`"tasks": {"tasks": [...]}` - has be double `"tasks"`, one for the 'file' label (`tasks.json`) and one for the tasks inside)
2. Open that file as your workspace.
3. Open the `Terminal` menu and select `Run Task...`
4. Observe that your tasks are available and run (or fail) as expected
5. Modify the file (e.g. add a new task configuration)
6. Repeat 3-4, and observe that the task list updated correctly.
---
1. Use task quick picks (commands `tasks configure`, `tasks run`) and confirm that they still work in each scope in single- and multi-root workspaces.
> NB: running very short-lived tasks (e.g. a `shell` task that just runs `echo`) can behave unexpectedly, with no terminal ever opening. This behavior is also found on `master`.

#### Changes

This PR has grown to touch quite a number of files, so it might be useful for reviewers if I explain why each change was necessary:
- Core
  - `preference-configurations.ts`: add an `isAnyConfig` convenience method equivalent to checking `isSectionName() || isConfigName()`
  - `preference-contribution.ts`: change schema update logic to respect the superset / subset relationships among scopes. See https://github.com/eclipse-theia/theia/pull/8917#issuecomment-767592610.
  - `preference-provider.ts`, `preference-service.ts`: add section name argument to `getConfigUri` and `getContainingConfigUri` to allow callers to find the URI of all preference files, not just `settings.json`. This is more systematic than the current implementation, and obviates workarounds using `PreferenceService.resolve` to find the URI for non-settings sections.
- Debug
  - `debug-schema-updater.ts`: add `launch` to the workspace file schema.
- Preferences
  - `folders-preference-provider.ts`, `user-configs-preference-provider.ts`, `workspace-preference-provider`: update `getConfigUri` and `getContainingConfigUri` to accommodate section name.
  - `section-preference-provider.ts`: minor cleanup to remove a non-null assertion.
  - `workspace-file-preference-provider.ts`: changes to handle sections other than `settings`, and to accommodate sections both inside the `settings` object and outside.
- Tasks
  - `quick-open-tasks.ts`: refactor the `configure` method to accommodate workspace-scoped tasks.
  - `task-configuration-manager.ts`: rewrite logic to accommodate workspace-scoped tasks.
  - `task-configuration-model.ts`: make scope protected because in the case of workspace scope it becomes unreliable (sometimes pointing to folder, sometimes pointing to workspace file). It wasn't being accessed, and the only way to get access to a model is to know the scope you're interested in, so you shouldn't have to ask the model for that information.
  - `task-configurations.ts`: change to allow configure to work in workspace scope.
  - `task-preferences.ts`: better default value for preferences.
  - `task-schema-updater.ts`: make the basic schema for tasks publicly accessible (it is for preferences) so that it can be used for any defaulting. Also add tasks the workspace-file schema.
- Workspace
  - `workspace-commands.ts`, `workspace-frontend-contribution.ts`: add a command to open the workspace file. It exists in VSCode and is parallel to commands to open e.g. the user-scope `tasks.json`.
  - `workspace-schema-updater.ts`, `workspace-frontend-module.ts`, `workspace-service.ts`: add a `WorkspaceSchemaUpdater` that can be accessed and updated by external packages (debug, tasks, preferences) to allow them to register their subschemas for the workspace file. Modify helper functions that assume that the schema will only ever include `settings`.

#### Review checklist

- [x] as an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

colin-grant-work commented:
@tsmaeder, I think you took the last thorough pass through the tasks system - I wonder if you'd be willing to take a look at this? I've made some changes to logic in the `TaskConfigurationManager` that may have some side effects that I haven't anticipated. In particular, there were some fallbacks to user scope if a find failed in folder scope, and I've removed that behavior since it seemed odd to tell a caller that tasks were in one scope when they were in another, but maybe that was hasty of me.

kenneth-marut-work reviewed (commented):
This seems to be working for me under the testing steps outlined and through experimenting with various other workspace configurations and setups. I haven't found any issues with your updates except that I'm not getting any sort of auto-suggest behavior inside the workspace file for `tasks`. Tried to track it down but was unsuccessful, not sure why that behavior isn't being enabled. 

Otherwise had some minor comments 👌

tsmaeder commented:
@colin-grant-work am I understanding correctly that this would also enable workspace-level launch configurations, not just tasks?

colin-grant-work commented:
@tsmaeder, workspace-scoped launch configurations are actually already supported, but they're formatted differently than VSCode expects. Instead of 
```
{
"settings": {...}
"launch": {"configurations": [...]}
}
```
The current system will do something like this:
```
{
"settings": {
  "launch": {...}
  }
}
```
This PR would adjust the way launch preferences are handled to agree with the format of VSCode's workspace file.

colin-grant-work commented on packages/workspace/src/browser/workspace-schema-updater.ts:
```diff
@@ -0,0 +1,115 @@
+/********************************************************************************
+ * Copyright (C) 2021 YourCompany and others.
```
Good catch. Fixed.

colin-grant-work commented on packages/task/src/browser/task-configuration-manager.ts:
```diff
@@ -71,14 +75,12 @@ export class TaskConfigurationManager {
 
-    protected readonly models = new Map<string, TaskConfigurationModel>();
-    protected userModel: TaskConfigurationModel;
+    protected readonly models = new Map<string | number, TaskConfigurationModel>();
```
👍

colin-grant-work commented on packages/preferences/src/browser/workspace-file-preference-provider.ts:
```diff
@@ -41,16 +41,33 @@ export class WorkspaceFilePreferenceProvider extends AbstractResourcePreferenceP
 
+    @postConstruct()
+    protected async init(): Promise<void> {
+        this.workspaceService.updateSchema('settings', {
```
Good call, didn't realize there was one, since the schema was specified separately for the workspace.

colin-grant-work commented:
@kenneth-marut-work, thanks for the review. I've implemented your suggestions and tracked down the bug with the auto-suggestion functionality.

>  I haven't found any issues with your updates except that I'm not getting any sort of auto-suggest behavior inside the workspace file for `tasks`. Tried to track it down but was unsuccessful, not sure why that behavior isn't being enabled.

Turned out there was a race condition in the schema update logic. I've implemented a queue to handle it, and you should see suggestions for `tasks`, `launch`, `folders` (if it isn't already present), and `settings`.

RomanNikitenko commented:
> workspace-scoped launch configurations are actually already supported, but they're formatted differently than VSCode expects. 

Looks like VS Code supports both formats
 
![settings_launch](https://user-images.githubusercontent.com/5676062/105350871-72235c80-5bf4-11eb-82de-649ae9c2d65c.gif)

colin-grant-work commented:
@RomanNikitenko, good catch. That's a bit annoying. The good news is (and please confirm that you also see this behavior), that you can do it only one way at a time. If there are `launch` configs **outside** settings, they take precedence over those **inside** settings:

![image](https://user-images.githubusercontent.com/62660806/105369521-42f6f600-5bc8-11eb-96f3-7c589358cdcb.png)

![image](https://user-images.githubusercontent.com/62660806/105369549-48ecd700-5bc8-11eb-8e7b-8acaa0684e6c.png)

Arranging for either...or is much easier than arranging for both...and. I'll modify the code accordingly.

colin-grant-work commented:
I've pushed code that should replicate the behavior seen in VSCode: it will read sections (`launch`, `tasks`) that exist inside `settings`, and if the user programmatically modifies those sections, it will perform the modifications inside `settings`. If it finds sections outside `settings`, it will favor that data over data inside `settings`, and any programmatic modifications will target the independent section.

RomanNikitenko reviewed (commented):
I tested a little the changes and can confirm - I'm able to run tasks configurations defined in the workspace configuration file - @colin-grant-work thanks a lot for such cool feature!

I faced with some issues within testing the PR, for example:
- I got the error ``` Global task cannot be customized ``` at attempt to apply `Configure Task` action to a workspace level task. You could try reproduce it by pressing the `gear` icon from `Terminal => Run task` menu
- Looks like workspace level tasks are not available from `Terminal => Configure Tasks` menu

@colin-grant-work 
could you take a look when you have a chance

RomanNikitenko commented:
I noticed some strange behavior related to launch configurations.

Steps to reproduce:
- multi-root project with 2 projects, I have `theia` and `vscode`, for example
- add a launch config on workspace level
- check that it's available for running for both projects from `Debug` panel
- add another configuration to `theia/.theia/launch.json` file
- go to the `Debug` panel again and check if the `workspace` level config is still available for running

As result I can see that the `workspace` launch config is available for running only for one project.

colin-grant-work commented:
> I noticed some strange behavior related to launch configurations.
> ...

I'm seeing the same behavior on `master`. I'll check it out a bit, but if it's going to take much messing with the debug-internal logic, I'm probably inclined to keep it separate from this PR.

RomanNikitenko commented:
> @RomanNikitenko, good catch. That's a bit annoying. The good news is (and please confirm that you also see this behavior), that you can do it only one way at a time. If there are launch configs outside settings, they take precedence over those inside settings:

Yes, I see the same behavior that you mentioned.
I added comment about it here: https://github.com/eclipse-theia/theia/issues/8836#issuecomment-766930716

colin-grant-work commented:
@danarad05, I'm having some trouble, and it looks like you're the most recent person to touch the code. It looks like #8761 added this code to the `PreferenceSchemaProvider`:

https://github.com/eclipse-theia/theia/blob/42042974c59f148974de5e8fd90cba7222cee82f/packages/core/src/browser/preferences/preference-contribution.ts#L393-L404

The `tasks` preferences are added in `'resource'` (i.e. `PreferenceScope.Folder`) scope, and so they are not being added to the workspace-scoped schema. That means that I can get auto-complete help if I put `tasks` outside the `settings` object in the workspace file, but not inside `settings`, because `tasks` isn't in the `settings` schema for workspace files. Did you have a strategy in mind for preferences that should be added to both scopes, or should I cook one up?

danarad05 commented:
> Did you have a strategy in mind for preferences that should be added to both scopes, or should I cook one up
@colin-grant-work 
I apologize for mishap, if you could fix it that would be great, thanks.

colin-grant-work commented:
@danarad05, I've modified the update and remove methods as follows:

```typescript
    protected updateSchemaProps(key: string, property: PreferenceDataProperty): void {
        this.combinedSchema.properties[key] = property;

        switch (property.scope) {
            case PreferenceScope.Folder:
                this.folderSchema.properties[key] = property;
            // Fall through. isValidInScope implies that User ⊃ Workspace ⊃ Folder,
            // so anything we add to folder should be added to workspace, but not vice versa.
            case PreferenceScope.Workspace:
                this.workspaceSchema.properties[key] = property;
                break;
        }
    }

    protected removePropFromSchemas(key: string): void {
        // If we remove a key from combined, it should also be removed from all narrower scopes.
        delete this.combinedSchema.properties[key];
        delete this.workspaceSchema.properties[key];
        delete this.folderSchema.properties[key];
    }
```

Does that seem reasonable?

colin-grant-work commented:
@RomanNikitenko, I think I've addressed your concerns about the `Configure` actions. I took a pass through the code and addressed that and any other cases of special-casing the `Global` scope. That's led to some refactoring in the `quick-open-tasks.ts` file. I also added a workspace command to open a workspace configuration file (if one exists) to parallel the `Open User Tasks` command, since VSCode also has an `Open Workspace Configuration File` command.

@kenneth-marut-work, re: your checks on the completion suggestion, I've made some changes in `PreferenceSchemaProvider` that should mean that you get auto-complete both inside the `settings` object in a workspace configuration file, and outside it as an independent `tasks` object.

danarad05 commented:
> @danarad05, I've modified the update and remove methods as follows:
> 
> ```ts
>     protected updateSchemaProps(key: string, property: PreferenceDataProperty): void {
>         this.combinedSchema.properties[key] = property;
> 
>         switch (property.scope) {
>             case PreferenceScope.Folder:
>                 this.folderSchema.properties[key] = property;
>             // Fall through. isValidInScope implies that User ⊃ Workspace ⊃ Folder,
>             // so anything we add to folder should be added to workspace, but not vice versa.
>             case PreferenceScope.Workspace:
>                 this.workspaceSchema.properties[key] = property;
>                 break;
>         }
>     }
> 
>     protected removePropFromSchemas(key: string): void {
>         // If we remove a key from combined, it should also be removed from all narrower scopes.
>         delete this.combinedSchema.properties[key];
>         delete this.workspaceSchema.properties[key];
>         delete this.folderSchema.properties[key];
>     }
> ```
> 
> Does that seem reasonable?

yep, thanks

kenneth-marut-work commented:
> @kenneth-marut-work, re: your checks on the auto-completion suggestions, I've made some changes in `PreferenceSchemaProvider` that should mean that you get auto-complete both inside the `settings` object in a workspace configuration file, and outside it as an independent `tasks` object.

I've confirmed that the auto-completion is working for both scenarios. Nice work.

I did however come across a new issue when using the `Configure Tasks` command. If I start with a new, empty workspace and select `Configure Tasks` -> `Open or create tasks.json file (Workspace)`, the file isn't opened until I run the same command again, and when it *is* opened, the file has a strange default value:

![image](https://user-images.githubusercontent.com/62660714/106018042-c01de180-6086-11eb-8a1a-6c652229c537.png)

This seems to be the case for any of the presets selected

RomanNikitenko commented on packages/workspace/src/browser/workspace-frontend-contribution.ts:
```diff
@@ -153,6 +153,14 @@ export class WorkspaceFrontendContribution implements CommandContribution, Keybi
+        commands.registerCommand(WorkspaceCommands.OPEN_WORKSPACE_FILE, {
+            isEnabled: () => this.workspaceService.saved,
+            execute: () => {
+                if (this.workspaceService.saved && this.workspaceService.workspace) {
```
Looks like the check for ` this.workspaceService.workspace` is redutdant: `this.workspaceService.saved` contains this check https://github.com/eclipse-theia/theia/blob/85aa2d5dba8f9b674d896c300f7d3a36016a7e00/packages/workspace/src/browser/workspace-service.ts#L524

colin-grant-work replied:
Fair point. I think I had them in the other order originally, but that wouldn't make it less redundant 😄

colin-grant-work replied:
Actually it turns out it was a typeguard. Without that check, Typescript isn't sure that `this.workspacesService.workspace.resource` (the URI for the workspace file) exists, so I'm going to leave it in.

RomanNikitenko commented:
> @RomanNikitenko, I think I've addressed your concerns about the Configure actions. I took a pass through the code and addressed that and any other cases of special-casing the Global scope. That's led to some refactoring in the quick-open-tasks.ts file. I also added a workspace command to open a workspace configuration file (if one exists) to parallel the Open User Tasks command, since VSCode also has an Open Workspace Configuration File command.

I tested after PR update and I didn't face any critical issue except the problem which @kenneth-marut-work posted above - I reproduced it.

Another minor issue is related to task schema support for Workspace-scoped tasks:  task system suggests me all known types when I'm creating a new task on Workspace level. AFAIK it should be only `shell` or `process` task type for Workspace scoped tasks. But it's OK for me to create a separate issue for that problem.

colin-grant-work commented:
@kenneth-marut-work, thanks for taking another look. I think I've addressed the issue of the strange opening behavior. A funky workaround had been employed to trick the `PreferenceService` into revealing the URI of the file that needed to be made or modified, and that workaround involved setting `tasks` to `{}`. (the empty string  in `"": {}` was my fault). Instead of that workaround, I've modified `PreferenceService.getConfigUri` to accept a `sectionName` argument (e.g. `tasks`) so that you can retrieve the URI of any preference file directly.

colin-grant-work commented:
@RomanNikitenko 
> Another issue (minor) is related to task schema support for Workspace-scoped tasks: task system suggests me all known types when I'm creating a new task on Workspace level. AFAIK it should be only `shell` or `process` task type for Workspace scoped tasks. But it's OK for me to create a separate issue for that problem.

That sounds reasonable, and it'll mean modifying the `TasksSchemaUpdater` to maintain multiple schemas, just like the `PreferenceSchemaProvider` maintains user-, workspace-, and folder-scoped schemas. If you're willing to defer that to another issue, I'd prefer that, as this one is already touching quite a bit.

RomanNikitenko commented:
@colin-grant-work 
thanks for the update - I'm going to test the changes ASAP.

About the task type limitation for `User` level and `Workspace` scoped tasks - I created the separate issue: https://github.com/eclipse-theia/theia/issues/9009

Please feel free to edit it or add other required info.
thanks!

RomanNikitenko reviewed (approved):
I tested the following for a multi-root workspace:
-  `Open Workspace Configuration File` command
- adding a task to the workspace config file
- task schema support for `Workspace` scoped tasks
- `Problems` view when there is an error in a `Workspace` scoped task configuration
- `Configure Task` action
- `Configure Task` => `Configure new task`
- running a workspace task
- terminate a workspace task
- restart a workspace task
- modify a workspace task configuration and check if the changes is applied for running

I also tested `launch` configs related logic:
- adding a launch configuration inside and outside of `settings` field of the workspace config file 
- schema support for launch configurations at a configuration creation

It works well for me, except https://github.com/eclipse-theia/theia/issues/8987 and https://github.com/eclipse-theia/theia/issues/9009. But it's out of the scope of the issue.

The PR also fixes https://github.com/eclipse-theia/theia/issues/8991

@colin-grant-work thanks a lot!

RomanNikitenko commented:
@kenneth-marut-work
could you review again when you have a chance https://github.com/eclipse-theia/theia/pull/8917#issuecomment-769425396
thanks in advance!

kenneth-marut-work commented:
@colin-grant-work 
Looks great, tested again and could not find any issues other than those @RomanNikitenko mentioned. I also confirmed that the empty string issue `"": {}` is gone.

Thanks for the thorough description of changes above. Ended up touching a lot of files, but excited for this new addition.
LGTM, and don't forget the changelog entry

colin-grant-work commented:
@vince-fugnitto, @marechal-p, would one of you be willing to take a look at this to make sure I haven't made any *unwise* changes to API? The changes to the `getConfigUri` method, `WorkspaceService` helper methods, and `WorkspaceFilePreferenceProvider` should definitely go in the `CHANGELOG` as (potentially) breaking, and I'm happy to add them, but it would be good to a checkon the code with an eye to larger code base, as well.

alvsan09 commented on packages/task/src/browser/quick-open-task.ts:
```diff
@@ -234,77 +234,41 @@ export class QuickOpenTask implements QuickOpenModel, QuickOpenHandler {
+        const scopes: TaskConfigurationScope[] = [TaskScope.Global];
+        if (this.workspaceService.opened) {
+            scopes.push(TaskScope.Workspace);
+            if (this.workspaceService.isMultiRootWorkspaceOpened) {
```
We are adding root folders only for multRootWorkSpaces but it the project is not multi root, we miss the entries so provided tasks are not listed for not multi root projects.

colin-grant-work replied:
Yes, good point. I changed the configured task system to treat a single folder root as 'workspace' scope, but provided tasks still treat them as folder scope. Would you prefer that the provided task system treat a single root as workspace scope, or adding folder scope even for single-root workspaces?

colin-grant-work replied:
It looks like VSCode's approach is to favor the folder interpretation, so I'll modify the logic to add workspace scope only if the workspace is saved.

alvsan09 replied:
The new version looks fine to me, I tried it and it works well !

alvsan09 commented on packages/task/src/browser/quick-open-task.ts:
```diff
@@ -234,77 +234,41 @@ export class QuickOpenTask implements QuickOpenModel, QuickOpenHandler {
-                        return item;
-                    })
-                );
+        let isFirstGroup = true;
```
I am wondering if we should replace this flag with this.items.length instead.

colin-grant-work replied:
I agree it's a bit unnecessary, but I think I prefer to keep the flag, because if someone later refactored to build the `items` array in the function scope and then assign it to `this.items` at the end, they could easily miss the side effect that that would have on the border display.

alvsan09 replied:
It's ok to keep it then !, thanks !

alvsan09 reviewed (changes requested):
It looks pretty good, I just spotted a regression that should be easy to fix.

colin-grant-work commented:
If there are no objections, I'll merge this tomorrow morning. @alvsan09, are you happy with it as is?

alvsan09 commented:
Hi @colin-grant-work , 
Yes I am fine with the updates, and it looks good to be merged !
thanks !
Regards

RomanNikitenko commented on packages/task/src/browser/task-configuration-manager.ts:
```diff
@@ -204,6 +219,26 @@ export class TaskConfigurationManager {
+    protected readonly toDisposeOnDelegateChange = new DisposableCollection();
+    protected updateWorkspaceModel(): void {
+        const newDelegate = this.workspaceService.saved ? this.workspacePreferences : this.folderPreferences;
+        const effectiveScope = this.workspaceService.saved ? TaskScope.Workspace : this.workspaceService.tryGetRoots()[0]?.resource.toString();
```
@colin-grant-work 
Hello!
Could you share what use case here is handled.

I thought that the use case is `Untittled (Workspace)` - so unsaved:
![untitled](https://user-images.githubusercontent.com/5676062/115567997-872b8c80-a2c4-11eb-9db9-352cfcd9e288.png)

but I get  `newDelegate = this.workspacePreferences` and `effectiveScope = TaskScope.Workspace` for such use case.

What steps I should do to get:
- newDelegate = this.folderPreferences;
- effectiveScope = this.workspaceService.tryGetRoots()[0]?.resource.toString();
 
Thanks in advance!

colin-grant-work replied:
Hey, @RomanNikitenko! As soon as a workspace file exists, the workspace is considered `.saved`, even if you haven't officially named it and saved it yourself. The case in which the delegate is folder preference is in a single-root unsaved workspace - i.e. the case in which the user uses 'Open folder...' to open a single folder as the workspace.

There was a slight hiccup with this logic where the `TaskConfigurationManager` wasn't switching to the workspace preferences delegate as soon as a workspace file was created, but it that was fixed recently in #9331. It looks like you're seeing that change in action: as soon as you have a workspace file, whether you've titled it or not, the `TaskConfigurationManager` starts considering the workspace file to be the source of information for workspace-scoped tasks.

RomanNikitenko replied:
thank you for the quick answer!
one more question:
> The case in which the delegate is folder preference is in a single-root unsaved workspace - i.e. the case in which the user uses 'Open folder...' to open a single folder as the workspace.

If user has a single folder - then tasks are considered on the folder level and should be handled [here](https://github.com/eclipse-theia/theia/blob/7d3b1d3b33cb565ec07965e5baf10c926006cdfd/packages/task/src/browser/task-configuration-manager.ts#L124)

So, I still don't see what use case we handle here.
I'm not saying that there is a problem - I'm investigating some area related to multi-root workspace and have found this place - so I just want to clarify it for myself.
thanks!

colin-grant-work replied:
You're right that all folder-scoped tasks, in both saved and unsaved workspaces, are handled by the folder preference provider. In the case of a single-root unsaved workspace, workspace scope *also* points to folder scope for the single folder. This parallels the preference system, which also collapses folder and workspace scope in the case of a single-root unsaved workspace. Since I'm not sure what your use case is, I'm not sure what steps to provide to 'reproduce' any particular behavior, but if you want to see how it works, you can do something like this:

1. Create a command that does something like this:
```typescript
{
    execute: () => {
        console.log("User scope:", this.taskConfigurationManager.getTasks(TaskScope.Global));
        console.log("Workspace scope:", this.taskConfigurationManager.getTasks(TaskScope.Workspace));
        for (const root of this.workspaceService.tryGetRoots()) {
            console.log(`Folder scope (${root.resource.path.base}):`, this.taskConfigurationManager.getTasks(root.resource.toString()));
        }
    }
}
```
Try that command in different kinds of workspaces. In a single-root unsaved workspace, the logs for workspace and folder scope should be the same (same `_scope` fields, too). In a saved workspace, they should be different. There aren't very many services that interface directly with the `TaskConfigurationManager` - mostly its output gets filtered through the `TaskService`, which is able to identify the fact that workspace scope and folder scope are duplicated in a single-root workspace (because the `label` and `_scope` fields are identical), so consumers of the data end up seeing what they should see: one set of folder-scoped tasks.

Conceivably, in a single-root unsaved workspace, we could just return an empty array for workspace tasks, but I prefer to keep the behavior of the various preference-related functionality consistent. Since the preferences service collapses workspace scope and folder scope in single-root workspaces, I think tasks (and `launch`), should do so, as well.

RomanNikitenko replied:
thank you for the detailed answer.

I think I was confused for the single-root use case(user have opened just a single folder) because
**before your explanation I thought** that:
- for the single-root use case tasks should be handled [here](https://github.com/eclipse-theia/theia/blob/7d3b1d3b33cb565ec07965e5baf10c926006cdfd/packages/task/src/browser/task-configuration-manager.ts#L124) on the `Folder` level - there is no need to create `TaskConfigurationModel` for the `Workspace` scope. 
- there should be no `Workspace` scoped tasks for the such use case - as they are duplicated - in the example that you provided above `taskConfigurationManager` returns the same tasks for the `Workspace` scope and for the `Folder` level

Thanks again for your time and efforts - now it's more clear for me!
